src/PSRule.Rules.Azure/Data/Policy/PolicyAssignmentVisitor.cs

```diff
@@ -710,7 +734,36 @@ protected virtual bool TryPolicyDefinition(PolicyAssignmentContext context, JObj
                 foreach (var parameter in parameters.Properties())
                     context.SetDefinitionParameterAssignment(result, parameter);
 
-                context.DefinitionIds.Add(policyDefinitionId, result);
+                // Check if definition with same parameters has already been added
+                if (context.DefinitionIds.ContainsKey(policyDefinitionId))
+                {
+                    var previousDefinition = context.DefinitionIds[policyDefinitionId];
+
+                    var foundDuplicateDefinition = true;
+                    foreach (var currentParameter in result.Parameters)
+                    {
+                        if (previousDefinition.TryParameter(currentParameter.Key, out var previousParameterValue))
+                        {
+                            if (!ParametersEqual(context, previousParameterValue, currentParameter.Value))
+                            {
+                                foundDuplicateDefinition = false;
+                                break;
+                            }
+                        }
+                    }
+
+                    // Overwrite definition if we found differences
+                    if (!foundDuplicateDefinition)
+                        context.DefinitionIds[policyDefinitionId] = result;
+
+                    // Otherwise skip adding definition
+                    else
+                        return false;
+                }
+
+                // Add definition for the first time
+                else
+                    context.DefinitionIds.Add(policyDefinitionId, result);
```

Comment:
@ArmaanMcleod Do we even need to store the policy definitions in `context.DefinitionIds` as a dictionary? Couldn't we just care about the current policy definition we are processing only.

The only real reference to this is: https://vscode.dev/github/armaanmcleod/psrule.rules.azure/blob/607b51868b45767215d852c2bb1644151e2faf5c/src/PSRule.Rules.Azure/Data/Policy/PolicyAssignmentVisitor.cs#L584-L589

After we've processed a policy definition we move on to the next one. Sure, if we reprocess the same definition there might be a little more processing for same policy + same parameters. But storing all policies in memory, also has a hit.

Reply:
@BernieWhite I think we could just store the definition parameters, that would probably reduce memory footprint alot.

I probably left it like this when I was getting inspiration from the `TemplateVisitor` code 😄.